### PR TITLE
add decision placeholder controller

### PR DESCRIPTION
**Pull request by elgnay**
The related task: https://github.com/open-cluster-management/backlog/issues/12421

Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/controllers/scheduling/placeholder_controller.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	"k8s.io/klog/v2"
+
+	clusterclient "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+	clusterinformerv1alpha1 "github.com/open-cluster-management/api/client/cluster/informers/externalversions/cluster/v1alpha1"
+	clusterlisterv1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	clusterlisterv1alpha1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1alpha1"
+	clusterapiv1alpha1 "github.com/open-cluster-management/api/cluster/v1alpha1"
+)
+
+// decisionPlaceholderController creates empty decisions as placeholders in status of PlacementDecisions
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/controllers/scheduling/placeholder_controller.go]: Done.

File: pkg/controllers/scheduling/placeholder_controller.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	// no work if PlacementDecision has not been created yet
+	if len(placementDecisions) == 0 {
+		return nil
+	}
+
+	if placement.Spec.NumberOfClusters == nil {
+		// TODO: support placement without Spec.NumberOfCluster
+		klog.Warningf("Placement without Spec.NumberOfCluster is not supported yet: %q", queueKey)
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/controllers/scheduling/placeholder_controller.go]: Removed.

File: pkg/controllers/scheduling/placeholder_controller.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		return fmt.Errorf("multiple placementdecisions is not supported yet")
+	}
+
+	placementDecision := placementDecisions[0]
+	newPlacementDecision := placementDecision.DeepCopy()
+	decisions := newPlacementDecision.Status.Decisions
+	decisionNum := len(decisions)
+
+	// truncate decision slice if the number of desisions is larger than the desired NOC
+	if decisionNum > desiredNOD {
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/controllers/scheduling/placeholder_controller.go]: Named updated.

File: pkg/controllers/scheduling/placeholder_controller.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+// or creating a separate one.
+type decisionPlaceholderController struct {
+	clusterClient           clusterclient.Interface
+	clusterLister           clusterlisterv1.ManagedClusterLister
+	placementLister         clusterlisterv1alpha1.PlacementLister
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister
+}
+
+// NewDecisionPlaceholderController return an instance of decisionPlaceholderController
+func NewDecisionPlaceholderController(
```
Comment (deads2k) [inline pkg/controllers/scheduling/placeholder_controller.go:36]: I missed where in this controller we build the list of clusters.

Comment (elgnay) [reply pkg/controllers/scheduling/placeholder_controller.go:36]: This controller does not list clusters. In func updateClusterDecisions(), it creates empty cluster decisions as placeholders in the status of placementdecisions, and guarantees each placementdecision contains the correct number of cluster decisions (some of them might be empty). Another controller (created in PR https://github.com/open-cluster-management/placement/pull/7) will monitor placementdecisions and list clusters to pick up feasible ones for the unscheduled cluster decisions (the empty ones).

File: pkg/controllers/scheduling/placeholder_controller.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		}, placementDecisionInformer.Informer()).
+		WithSync(c.sync).
+		ToController("DecisionPlaceholderController", recorder)
+}
+
+func (c *decisionPlaceholderController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		// ignore placement whose key is not in format: namespace/name
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/controllers/scheduling/placeholder_controller.go:79]: Code updated to handle error with the recommended function.

**Comment by elgnay** [comment]
Based on discussion on https://github.com/open-cluster-management/placement/pull/7, this PR is no longer needed.

**Pull request closed without merging**